File: ambari-web/app/models/alert.js
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var App = require('app');

/**
 * Defines structure for App.Alert class. Keys mentioned here are for JSON data
 * which comes back from NAGIOS server.
 */
App.Alert = Em.Object.extend({
  title: null,//service_description in ajax response
  serviceType: null,
  status: null,//current_state in ajax response
  message: null,//plugin_output in ajax response
  hostName: null,
  lastCheck: null,
  lastTime: null,

  date: function () {
    return DS.attr.transforms.date.from(this.get('lastTime'));
  }.property('lastTime'),

  /**
   * Used to show correct icon in UI
   */
  isOk: function () {
    return this.get('status') == "0";
  }.property('status'),

  /**
   * Used to show correct icon in UI
   */
  isWarning: function () {
    return this.get('status') == "1";
  }.property('status'),

  /**
   * Used to show correct icon in UI
   */
  isCritical: function() {
    return this.get('status') == '2';
  }.property('status'),

  /**
   * Used to show correct icon in UI
   */
  isPassive: function() {
    return this.get('status') == '3';
  }.property('status'),

  /**
   * Used to show only required alerts at the service level
   */
  ignoredForServices: function() {
    return ['NodeManager health', 'NodeManager process', 'TaskTracker process', 'RegionServer process', 'DataNode process', 'DataNode space', 'ZooKeeper Server process', 'Supervisors process'].contains(this.get('title'));
  }.property('title'),

  /**
   * Used to show only required alerts at the host level
   */
  ignoredForHosts: function() {
    return this.get('title').indexOf('Percent') != -1;
  }.property('title'),

  /**
   * Provides how long ago this alert happened.
   * 
   * @type {String}
   */
  timeSinceAlert: function () {
    var d = this.get('date');
    if (d) {
      var timeFormat = this.t('services.alerts.OK.timePrefix');
      switch (this.get('status')) {
        case "1":
          timeFormat = this.t('services.alerts.WARN.timePrefix');
          break;
        case "2":
          timeFormat = this.t('services.alerts.CRIT.timePrefix');
          break;
        case "3":
          timeFormat = this.t('services.alerts.MAINT.timePrefix');
          break;
        case "4":
          timeFormat = this.t('services.alerts.UNKNOWN.timePrefix');
          break;
      }
      var prevSuffix = $.timeago.settings.strings.suffixAgo;
      $.timeago.settings.strings.suffixAgo = '';
      var since = timeFormat.format($.timeago(this.makeTimeAtleastMinuteAgo(d)));
      $.timeago.settings.strings.suffixAgo = prevSuffix;
      return since;
    } else if (d == 0) {
      var timeFormat = this.t('services.alerts.OK.timePrefixShort');
      switch (this.get('status')) {
        case "1":
          timeFormat = this.t('services.alerts.WARN.timePrefixShort');
          break;
        case "2":
          timeFormat = this.t('services.alerts.CRIT.timePrefixShort');
          break;
        case "3":
          timeFormat = this.t('services.alerts.MAINT.timePrefixShort');
          break;
        case "4":
          timeFormat = this.t('services.alerts.UNKNOWN.timePrefixShort');
          break;
      }
      return timeFormat;
    } else {
      return "";
    }
  }.property('date', 'status'),
  
  makeTimeAtleastMinuteAgo: function(d){
    var time = d.getTime();
    var diff = App.dateTime() - time;
    if (diff < 60000) {
      diff = 60000 - diff;
      return new Date(time - diff);
    }
    return d;
  },

  /**
   * Provides more details about when this alert happened.
   * 
   * @type {String}
   */
  timeSinceAlertDetails: function () {
    var details = "";
    var date = this.get('date');
    if (date) {
      var dateString = date.toDateString();
      dateString = dateString.substr(dateString.indexOf(" ") + 1);
      dateString = Em.I18n.t('services.alerts.occurredOn').format(dateString, date.toLocaleTimeString());
      details += dateString;
    }
    var lastCheck = this.get('lastCheck');
    if (lastCheck) {
      lastCheck = new Date(lastCheck * 1000);
      details = details ? details + Em.I18n.t('services.alerts.brLastCheck').format($.timeago(lastCheck)) : Em.I18n.t('services.alerts.lastCheck').format($.timeago(lastCheck));
    }
    return details;
  }.property('lastCheck', 'date'),

  /**
   * Used to show appropriate service label in UI
   */
  serviceName: function () {
    if (this.get('serviceType')) {
      var type = this.get('serviceType').toLowerCase();
      switch (type) {
        case 'mapreduce':
          return 'MapReduce';
        case 'hdfs':
          return 'HDFS';
        case 'hbase':
          return "HBase";
        case 'zookeeper':
          return "Zookeeper";
        case 'oozie':
          return "Oozie";
        case 'hive':
          return 'Hive';
      }
    }
    return null;
  }.property('serviceType'),

  /**
   * Used to provide appropriate service link in UI
   */
  serviceLink: function () {
    if (this.get('serviceType')) {
      var type = this.get('serviceType').toLowerCase();
      switch (type) {
        case 'mapreduce':
          return '#/main/services/MAPREDUCE/summary';
        case 'hdfs':
          return '#/main/services/HDFS/summary';
        case 'hbase':
          return '#/main/services/HBASE/summary';
        case 'zookeeper':
          return '#/main/services/ZOOKEEPER/summary';
        case 'oozie':
          return '#/main/services/OOZIE/summary';
        case 'hive':
          return '#/main/services/HIVE/summary';
      }
    }
    return null;
  }.property('serviceType')
});

File: ambari-server/src/main/java/org/apache/ambari/server/state/Alert.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ambari.server.state;

import org.apache.commons.lang.StringUtils;
import org.codehaus.jackson.annotate.JsonProperty;
/**
 * An alert represents a problem or notice for a cluster.
 */
public class Alert {
  private String cluster = null;
  private String name = null;
  private String instance = null;
  private String service = null;
  private String component = null;
  private String hostName = null;
  private AlertState state = AlertState.UNKNOWN;
  private String label = null;
  private String text = null;
  private long timestamp = 0L;

  // Maximum string size for MySql TEXT (utf8) column data type
  protected final static int MAX_ALERT_TEXT_SIZE = 32617;


  /**
   * Constructor.
   * @param alertName the name of the alert
   * @param alertInstance instance specific information in the event that two alert
   *    types can be run, ie Flume.
   * @param serviceName the service
   * @param componentName the component
   * @param hostName the host
   * @param alertState the state of the alertable event
   */
  public Alert(String alertName, String alertInstance, String serviceName,
      String componentName,  String hostName, AlertState alertState) {
    name = alertName;
    instance = alertInstance;
    service = serviceName;
    component = componentName;
    this.hostName = hostName;
    state = alertState;
  }

  public Alert() {
  }

  /**
   * @return the name
   */

  @JsonProperty("name")
  public String getName() {
    return name;
  }

  /**
   * @return the service
   */
  @JsonProperty("service")
  public String getService() {
    return service;
  }

  /**
   * @return the component
   */
  @JsonProperty("component")
  public String getComponent() {
    return component;
  }

  /**
   * @return the host
   */
  @JsonProperty("host")
  public String getHostName() {
    return hostName;
  }

  /**
   * @return the state
   */
  @JsonProperty("state")
  public AlertState getState() {
    return state;
  }

  /**
   * @return a short descriptive label for the alert
   */
  @JsonProperty("label")
  public String getLabel() {
    return label;
  }

  /**
   * @param alertLabel a short descriptive label for the alert
   */
  @JsonProperty("label")
  public void setLabel(String alertLabel) {
    label = alertLabel;
  }

  /**
   * @return detail text about the alert
   */
  @JsonProperty("text")
  public String getText() {
    return text;
  }

  /**
   * @param alertText detail text about the alert
   */
  @JsonProperty("text")
  public void setText(String alertText) {
    // middle-ellipsize the text to reduce the size to 32617 characters
    text = StringUtils.abbreviateMiddle(alertText, "…", MAX_ALERT_TEXT_SIZE);
  }

  @JsonProperty("instance")
  public String getInstance() {
    return instance;
  }

  @JsonProperty("instance")
  public void setInstance(String instance) {
    this.instance = instance;
  }

  @JsonProperty("name")
  public void setName(String name) {
    this.name = name;
  }

  @JsonProperty("service")
  public void setService(String service) {
    this.service = service;
  }

  @JsonProperty("component")
  public void setComponent(String component) {
    this.component = component;
  }

  @JsonProperty("host")
  public void setHostName(String hostName) {
    this.hostName = hostName;
  }

  @JsonProperty("state")
  public void setState(AlertState state) {
    this.state = state;
  }

  @JsonProperty("timestamp")
  public void setTimestamp(long ts) {
    timestamp = ts;
  }

  @JsonProperty("timestamp")
  public long getTimestamp() {
    return timestamp;
  }

  /**
   * @return
   */
  @JsonProperty("cluster")
  public String getCluster() {
    return cluster;
  }

  @JsonProperty("cluster")
  public void setCluster(String cluster){
    this.cluster = cluster;
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public int hashCode() {
    final int prime = 31;
    int result = 1;
    result = prime * result + ((state == null) ? 0 : state.hashCode());
    result = prime * result + ((name == null) ? 0 : name.hashCode());
    result = prime * result + ((service == null) ? 0 : service.hashCode());
    result = prime * result + ((component == null) ? 0 : component.hashCode());
    result = prime * result + ((hostName == null) ? 0 : hostName.hashCode());
    result = prime * result + ((cluster == null) ? 0 : cluster.hashCode());
    result = prime * result + ((instance == null) ? 0 : instance.hashCode());
    return result;
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }

    if (obj == null) {
      return false;
    }

    if (getClass() != obj.getClass()) {
      return false;
    }

    Alert other = (Alert) obj;

    if (state != other.state) {
      return false;
    }

    if (name == null) {
      if (other.name != null) {
        return false;
      }
    } else if (!name.equals(other.name)) {
      return false;
    }

    if (service == null) {
      if (other.service != null) {
        return false;
      }
    } else if (!service.equals(other.service)) {
      return false;
    }

    if (component == null) {
      if (other.component != null) {
        return false;
      }
    } else if (!component.equals(other.component)) {
      return false;
    }

    if (hostName == null) {
      if (other.hostName != null) {
        return false;
      }
    } else if (!hostName.equals(other.hostName)) {
      return false;
    }

    if (cluster == null) {
      if (other.cluster != null) {
        return false;
      }
    } else if (!cluster.equals(other.cluster)) {
      return false;
    }

    if (instance == null) {
      if (other.instance != null) {
        return false;
      }
    } else if (!instance.equals(other.instance)) {
      return false;
    }


    return true;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append('{');
    sb.append("cluster=").append(cluster).append(", ");
    sb.append("state=").append(state).append(", ");
    sb.append("name=").append(name).append(", ");
    sb.append("service=").append(service).append(", ");
    sb.append("component=").append(component).append(", ");
    sb.append("host=").append(hostName).append(", ");
    sb.append("instance=").append(instance).append(", ");
    sb.append("text='").append(text).append("'");
    sb.append('}');
    return sb.toString();
  }
}
